File: scala/src/main/scala/tdigest/TDigest.scala
package tdigest

import collection.JavaConverters._
import scala.util.control.Breaks.{break, breakable}

class TDigest(private val maxSize: Long = 100) {
  var centroids = new java.util.ArrayList[Centroid](maxSize.toInt)
  private var sum: Double = 0.0
  private var count: Double = 0.0
  private var max: Double = Double.NaN
  private var min: Double = Double.NaN

  private def updateFrom(other: TDigest) = {
    this.centroids = other.centroids
    this.sum = other.sum
    this.count = other.count
    this.max = other.max
    this.min = other.min
  }

  def mergeUnsorted(unsortedValues: Iterable[Double]): TDigest = {
    val sorted: List[Double] = unsortedValues.toSeq.sorted.toList
    mergeSorted(sorted)
  }

  def mergeSorted(sortedValues: List[Double]): TDigest = {
    if (sortedValues.isEmpty) {
      return this.clone().asInstanceOf[TDigest]
    }

    val result = new TDigest(this.maxSize);
    result.count = this.count + sortedValues.length.toDouble

    val maybeMin: Double = sortedValues.head
    val maybeMax: Double = sortedValues.last

    if (this.count > 0.0) {
      result.min = this.min.min(maybeMin)
      result.max = this.max.max(maybeMax)
    } else {
      result.min = maybeMin
      result.max = maybeMax
    }

    val compressed = new java.util.ArrayList[Centroid](this.maxSize.toInt)

    var kLimit: Double = 1.0
    var qlimitTimesCount: Double = TDigest.kToQ(kLimit, this.maxSize.toDouble) * result.count
    kLimit += 1.0

    val centroidsIterator: scala.collection.BufferedIterator[Centroid] = this.centroids.iterator().asScala.buffered
    val sortedValuesIterator: scala.collection.BufferedIterator[Double] = sortedValues.iterator.buffered

    var curr: Centroid = if (centroidsIterator.hasNext) {
      val c: Centroid = centroidsIterator.head
      val curr2: Double = sortedValuesIterator.head
      if (c.getMean < curr2) {
        centroidsIterator.next().clone()
      } else {
        new Centroid(sortedValuesIterator.next(), 1.0)
      }
    } else {
      new Centroid(sortedValuesIterator.next(), 1.0)
    }

    var weightSoFar: Double = curr.getWeight

    var sumsToMerge: Double = 0.0
    var weightsToMerge: Double = 0.0

    while (centroidsIterator.hasNext || sortedValuesIterator.hasNext) {
      val next: Centroid = if (centroidsIterator.hasNext) {
        val c: Centroid = centroidsIterator.head
        if (!sortedValuesIterator.hasNext || c.getMean < sortedValuesIterator.head) {
          centroidsIterator.next().clone()
        } else {
          new Centroid(sortedValuesIterator.next(), 1.0)
        }
      } else {
        new Centroid(sortedValuesIterator.next(), 1.0)
      }

      val next_sum: Double = next.getMean * next.getWeight
      weightSoFar += next.getWeight

      if (weightSoFar <= qlimitTimesCount) {
        sumsToMerge += next_sum
        weightsToMerge += next.getWeight
      } else {
        result.sum = result.sum + curr.add(sumsToMerge, weightsToMerge)
        sumsToMerge = 0.0
        weightsToMerge = 0.0

        compressed.add(curr.clone());
        qlimitTimesCount = TDigest.kToQ(kLimit, this.maxSize.toDouble) * result.count
        kLimit += 1.0
        curr = next
      }
    }

    result.sum = result.sum + curr.add(sumsToMerge, weightsToMerge)
    compressed.add(curr);
    compressed.sort(Centroid.compare)

    result.centroids = compressed
    updateFrom(result)
    result
  }

  def mean(): Double = {
    if (count > 0.0) {
      sum / count
    } else {
      0.0
    }
  }

  /**
   * To estimate the value located at `q` quantile
   */
  def estimateQuantile(q: Double): Double = {
    if (centroids.isEmpty) {
      return 0.0
    }
    val rank: Double = q * count

    var pos: Int = 0
    var t: Double = 0d

    if (q > 0.5) {
      if (q >= 1.0) {
        return this.max
      }

      pos = 0
      t = count

      breakable {
        for ((centroid, k) <- centroids.asScala.zipWithIndex.reverse) {
          t -= centroid.getWeight
          if (rank >= t) {
            pos = k
            break
          }
        }
      }
    } else {
      if (q <= 0.0) {
        return this.min
      }

      pos = centroids.size() - 1
      t = 0.0

      breakable {
        for ((centroid, k) <- centroids.asScala.zipWithIndex) {
          if (rank < t + centroid.getWeight) {
            pos = k
            break
          }

          t += centroid.getWeight
        }
      }
    }

    var delta: Double = 0.0
    var minVal: Double = this.min
    var maxVal: Double = this.max

    if (centroids.size() > 1) {
      if (pos == 0) {
        delta = centroids.get(pos + 1).getMean - centroids.get(pos).getMean
        maxVal = centroids.get(pos + 1).getMean
      } else if (pos == (centroids.size() - 1)) {
        delta = centroids.get(pos).getMean - centroids.get(pos - 1).getMean
        minVal = centroids.get(pos - 1).getMean
      } else {
        delta = (centroids.get(pos + 1).getMean - centroids.get(pos - 1).getMean) / 2.0
        minVal = centroids.get(pos - 1).getMean
        maxVal = centroids.get(pos + 1).getMean
      }
    }

    val value = centroids.get(pos).getMean + ((rank - t) / centroids.get(pos).getWeight - 0.5) * delta
    TDigest.clamp(value, minVal, maxVal)
  }

}

object TDigest {
  private def kToQ(k: Double, d: Double): Double = {
    val k_div_d = k / d;
    if (k_div_d >= 0.5) {
      val base = 1.0 - k_div_d;
      1.0 - 2.0 * base * base
    } else {
      2.0 * k_div_d * k_div_d
    }
  }

  private def clamp(v: Double, lo: Double, hi: Double): Double = {
    if (v > hi) {
      hi
    } else if (v < lo) {
      lo
    } else {
      v
    }
  }

}

File: python/tdigest/tdigest.py
import numpy as np
import numba as nb

from .centroids import Centroids
from .centroid import Centroid
from . import centroid
from . import centroids


TDigest = np.dtype([
    ("centroids", centroids.Centroids),
    ("max_size", np.uint64),
    ("sum", np.float64),
    ("count", np.float64),
    ("max", np.float64),
    ("min", np.float64),
])


@nb.njit(cache=True)
def new() -> np.ndarray[TDigest]:
    out = np.zeros(1, dtype=TDigest)
    self = out[0]
    self["centroids"] = centroids.new()[0]
    self["max_size"] = centroids.capacity(self["centroids"])
    self["sum"] = np.float64(0.0)
    self["count"] = np.float64(0.0)
    self["max"] = np.nan
    self["min"] = np.nan
    return out


@nb.njit(cache=True)
def add_value(self: np.ndarray[TDigest], value: np.float64):
    add_sorted_values(self, np.array([value], dtype=np.float64))


@nb.njit(cache=True)
def add_sorted_values(self: np.ndarray[TDigest], values: np.ndarray[np.float64]) -> None:
    if values.size == 0:
        return
    result: np.ndarray[TDigest] = new()
    result[0]["count"] = self[0]["count"] + values.size
    maybe_min = values[0]
    maybe_max = values[-1]
    if self[0]["count"] > 0:
        # We know that min_ and max_ are numbers
        result[0]["min"] = np.minimum(self[0]["min"], maybe_min)
        result[0]["max"] = np.maximum(self[0]["max"], maybe_max)
    else:
        # We know that min_ and max_ are NaN.
        result[0]["min"] = maybe_min
        result[0]["max"] = maybe_max

    compressed_a: np.ndarray[Centroids] = centroids.new()
    compressed: Centroids = compressed_a[0]

    k_limit = np.float64(1.0)
    q_limit_times_count: np.float64 = k_to_q(k_limit, self[0]["max_size"]) * result[0]["count"]
    k_limit += 1

    centroid_index = np.uint64(0)
    values_index = np.uint64(0)

    cur: np.ndarray[Centroid] = np.zeros(1, dtype=Centroid)
    if (
        centroid_index < self[0]["centroids"]["size"]
        and self[0]["centroids"]["array"][centroid_index]["mean"] < values[values_index]
    ):
        cur = self[0]["centroids"]["array"][centroid_index:centroid_index+np.uint64(1)]
        centroid_index += np.uint64(1)
    else:
        cur = centroid.new(values[values_index], 1.0)
        values_index += np.uint64(1)

    weight_so_far = cur[0]["weight"]

    # Keep track of sums along the way to reduce expensive floating points
    sums_to_merge = np.float64(0.0)
    weights_to_merge = np.float64(0.0)

    next_: np.ndarray[Centroid] = np.zeros(1, dtype=Centroid)
    while (
        centroid_index < self[0]["centroids"]["size"]
        or values_index < values.size
    ):
        if (
            centroid_index < self[0]["centroids"]["size"]
            and (
                values_index >= values.size
                or self[0]["centroids"]["array"][centroid_index]["mean"] < values[values_index]
            )
        ):
            next_ = self[0]["centroids"]["array"][centroid_index:centroid_index+np.uint64(1)]
            centroid_index += np.uint64(1)
        else:
            next_ = centroid.new(values[values_index], 1.0)
            values_index += np.uint64(1)

        next_sum = next_[0]["mean"] * next_[0]["weight"]
        weight_so_far += next_[0]["weight"]

        if weight_so_far <= q_limit_times_count:
            sums_to_merge += next_sum
            weights_to_merge += next_[0]["weight"]
        else:
            result[0]["sum"] += centroid.add(cur[0], sums_to_merge, weights_to_merge)
            sums_to_merge = np.float64(0.0)
            weights_to_merge = np.float64(0.0)
            centroids.push_back(compressed, cur[0])
            q_limit_times_count = k_to_q(k_limit, self[0]["max_size"]) * result[0]["count"]
            k_limit += 1
            cur = next_

    result[0]["sum"] += centroid.add(cur[0], sums_to_merge, weights_to_merge)
    centroids.push_back(compressed, cur[0])
    centroids.sort(compressed)
    result[0]["centroids"] = compressed
    self[0] = result[0]


@nb.njit(cache=True)
def quantile(self: np.ndarray[TDigest], q: np.float64) -> np.float64:
    self: TDigest = self[0]
    if self["centroids"]["size"] == 0:
        return np.float64(0.0)
    rank: np.float64 = q * self["count"]
    pos = np.uint64(0)
    if q > np.float64(0.5):
        if q >= np.float64(1.0):
            return self["max"]
        pos = np.uint64(0)
        t = self["count"]
        for i in range(self["centroids"]["size"] - 1, -1, -1):
            c = self["centroids"]["array"][i]
            t -= c["weight"]
            if rank >= t:
                pos = np.uint64(i)
                break
    else:
        if q <= np.float64(0.0):
            return self["min"]
        pos = self["centroids"]["size"] - np.uint64(1)
        t = np.float64(0.0)
        for i in range(self["centroids"]["size"]):
            if rank < t + self["centroids"]["array"][i]["weight"]:
                pos = np.uint64(i)
                break
            t += self["centroids"]["array"][i]["weight"]

    delta = np.float64(0.0)
    min_ = self["min"]
    max_ = self["max"]
    if self["centroids"]["size"] > 1:
        if pos == np.uint64(0):
            delta = self["centroids"]["array"][pos + np.uint64(1)]["mean"] - self["centroids"]["array"][pos]["mean"]
            max_ = self["centroids"]["array"][pos + np.uint64(1)]["mean"]
        elif pos == self["centroids"]["size"] - np.uint64(1):
            delta = self["centroids"]["array"][pos]["mean"] - self["centroids"]["array"][pos - np.uint64(1)]["mean"]
            min_ = self["centroids"]["array"][pos - np.uint64(1)]["mean"]
        else:
            delta = (self["centroids"]["array"][pos + np.uint64(1)]["mean"] - self["centroids"]["array"][pos - np.uint64(1)]["mean"]) / 2
            min_ = self["centroids"]["array"][pos - np.uint64(1)]["mean"]
            max_ = self["centroids"]["array"][pos + np.uint64(1)]["mean"]

    c = self["centroids"]["array"][pos]
    value = c["mean"] + ((rank - t) / c["weight"] - np.float64(0.5)) * delta
    return clip(value, min_, max_)


@nb.njit(cache=True)
def k_to_q(k: np.float64, d: np.float64) -> np.float64:
    k_div_d: np.float64 = k / d
    if k_div_d >= np.float64(0.5):
        base: np.float64 = np.float64(1.0) - k_div_d
        return np.float64(1.0) - np.float64(2.0) * base * base
    else:
        return np.float64(2.0) * k_div_d * k_div_d


@nb.njit(cache=True)
def clip(x: np.float64, min_: np.float64, max_: np.float64) -> np.float64:
    if x < min_:
        return min_
    elif x > max_:
        return max_
    else:
        return x